### DISCUSSION TITLE
Bug and touch-ups for trivy v0.51.2 (with a proposed fix)

### DISCUSSION THREAD

**Discussion by bobsrac** [aquasecurity/trivy#6757] [Bugs]
### Description

Bug and touch-ups for trivy v0.51.2 (with a proposed fix)

* bug: Misconfigurations without a location cause a large error to be filed as a diagnostic
* clean-up: Change from '--scanners config' to '--scanners misconfig' to eliminate deprecation log message
* clean-up: Add '--quiet' flag to suppress log output, which is unused by trivy.lua

### Desired Behavior

Expected each misconfiguration from trivy to result in a well-formed diagnostic in neovim

Expected that Misconfigurations without location information would still show up as diagnostics, and they would default to the top of the file

### Actual Behavior

When I open the Dockerfile from "Reproduction Steps", I see a long error message from trivy as a diagnostic

The cause is in /lua/lint/linters/trivy.lua.  It creates a diagnostic reading StartLine and EndLine as integers without checking that they existed in the output from trivy.  My proposed fix is to:
1. check for all required fields from the parsed JSON needed to create a useful diagnostic from each Misconfiguration.
2. create defaults for col, end_col, lnum, end_lnum all initialized to 0, then override from StartLine and EndLine as before if those fields exist.

* nvim-lint/lua/lint/linters/trivy.lua does not check that all of the fields it uses exist in the JSON before creating and adding a diagnostic.  I've added that check to drop

Two other minor notes:
* trivy stdout is processed as a JSON, therefore the log lines are not useful.  trivy has a '--quiet' flag to suppress the log
* trivy complains the cli arg '--scanners config' is deprecated.  The updated arg: '--scanners misconfig'

Proposed trivy.lua:
$ cat trivy.lua
local severity_map = {
  ["LOW"] = vim.diagnostic.severity.INFO,
  ["MEDIUM"] = vim.diagnostic.severity.WARN,
  ["HIGH"] = vim.diagnostic.severity.ERROR,
}

return {
  cmd = "trivy",
  stdin = false,
  append_fname = true,
  args = { "--quiet", "--scanners", "misconfig", "--format", "json", "fs" },
  stream = "stdout",
  ignore_exitcode = false,
  parser = function(output, bufnr)
    local diagnostics = {}
    local ok, decoded = pcall(vim.json.decode, output)
    if not ok then
      return diagnostics
    end
    local fpath = vim.fn.fnamemodify(vim.api.nvim_buf_get_name(bufnr), ":t")
    for _, result in ipairs(decoded and decoded.Results or {}) do
      if result.Target == fpath then
        for _, misconfig in ipairs(result.Misconfigurations or {}) do
          if
            misconfig.Title
            and misconfig.Description
            and misconfig.CauseMetadata
            and misconfig.ID
            and misconfig.Severity
          then
            local col = 0
            local lnum = 0
            if misconfig.CauseMetadata.StartLine then
              col = misconfig.CauseMetadata.StartLine
              lnum = misconfig.CauseMetadata.StartLine - 1
            end

            local end_col = 0
            local end_lnum = 0
            if misconfig.CauseMetadata.EndLine then
              end_col = misconfig.CauseMetadata.EndLine
              end_lnum = misconfig.CauseMetadata.EndLine - 1
            end

            local err = {
              source = "trivy",
              message = string.format("%s %s", misconfig.Title, misconfig.Description),
              col = col,
              end_col = enc_col,
              lnum = lnum,
              end_lnum = end_lnum,
              code = misconfig.ID,
              severity = severity_map[misconfig.Severity],
            }
            table.insert(diagnostics, err)
          end
        end
      end
    end
    return diagnostics
  end,
}

### Reproduction Steps

```bash
1. Install trivy v0.51.2
2. Open a dockerfile that does not haev a HEALTHCHECK instruction -- this will yield a Misconfiguration whose CauseMetadata lacks "StartLine" and "EndLine" fields
3. View diagnostics for the buffer and note the entire stdout from trivy appears as a diagnostic prepended with following:

Parser Failed. Error message:
<path_to_trivy.lua>:29: attempt to perform arithmetic on field 'StartLine' (a nil value)

Output from linter:
...
```

### Target

Container Image

### Scanner

Misconfiguration

### Output Format

JSON

### Mode

Standalone

### Debug Output

```bash
Error is in the lua script.  stdout from trivy is not in error.  The --debug produces no change in output.  I've pasted the Dockerfile, and the output from trivy both before and after my patch:

$ cat Dockerfile
FROM ubuntu:24.04
LABEL maintainer="Matt Mahin <matt.mahin@gmail.com"

$ trivy --scanners config --format json fs Dockerfile
2024-05-22T23:01:32-06:00       WARN    '--scanners config' is deprecated. Use '--scanners misconfig' instead. See https://github.com/aquasecurity/trivy/discussions/5586 for the detail.
2024-05-22T23:01:32-06:00       INFO    Misconfiguration scanning is enabled
2024-05-22T23:01:32-06:00       INFO    Detected config files   num=1
{
  "SchemaVersion": 2,
  "CreatedAt": "2024-05-22T23:01:32.679931895-06:00",
  "ArtifactName": "Dockerfile",
  "ArtifactType": "filesystem",
  "Metadata": {
    "ImageConfig": {
      "architecture": "",
      "created": "0001-01-01T00:00:00Z",
      "os": "",
      "rootfs": {
        "type": "",
        "diff_ids": null
      },
      "config": {}
    }
  },
  "Results": [
    {
      "Target": "Dockerfile",
      "Class": "config",
      "Type": "dockerfile",
      "MisconfSummary": {
        "Successes": 25,
        "Failures": 2,
        "Exceptions": 0
      },
      "Misconfigurations": [
        {
          "Type": "Dockerfile Security Check",
          "ID": "DS002",
          "AVDID": "AVD-DS-0002",
          "Title": "Image user should not be 'root'",
          "Description": "Running containers with 'root' user can lead to a container escape situation. It is a best practice to run containers as non-root users, which can be done by adding a 'USER' statement to the Dockerfile.",
          "Message": "Specify at least 1 USER command in Dockerfile with non-root user as argument",
          "Namespace": "builtin.dockerfile.DS002",
          "Query": "data.builtin.dockerfile.DS002.deny",
          "Resolution": "Add 'USER \u003cnon root user name\u003e' line to the Dockerfile",
          "Severity": "HIGH",
          "PrimaryURL": "https://avd.aquasec.com/misconfig/ds002",
          "References": [
            "https://docs.docker.com/develop/develop-images/dockerfile_best-practices/",
            "https://avd.aquasec.com/misconfig/ds002"
          ],
          "Status": "FAIL",
          "Layer": {},
          "CauseMetadata": {
            "Provider": "Dockerfile",
            "Service": "general",
            "Code": {
              "Lines": null
            }
          }
        },
        {
          "Type": "Dockerfile Security Check",
          "ID": "DS026",
          "AVDID": "AVD-DS-0026",
          "Title": "No HEALTHCHECK defined",
          "Description": "You should add HEALTHCHECK instruction in your docker container images to perform the health check on running containers.",
          "Message": "Add HEALTHCHECK instruction in your Dockerfile",
          "Namespace": "builtin.dockerfile.DS026",
          "Query": "data.builtin.dockerfile.DS026.deny",
          "Resolution": "Add HEALTHCHECK instruction in Dockerfile",
          "Severity": "LOW",
          "PrimaryURL": "https://avd.aquasec.com/misconfig/ds026",
          "References": [
            "https://blog.aquasec.com/docker-security-best-practices",
            "https://avd.aquasec.com/misconfig/ds026"
          ],
          "Status": "FAIL",
          "Layer": {},
          "CauseMetadata": {
            "Provider": "Dockerfile",
            "Service": "general",
            "Code": {
              "Lines": null
            }
          }
        }
      ]
    }
  ]
}

$ trivy --quiet --scanners misconfig --format json fs Dockerfile
{
  "SchemaVersion": 2,
  "CreatedAt": "2024-05-22T23:02:08.981961151-06:00",
  "ArtifactName": "Dockerfile",
  "ArtifactType": "filesystem",
  "Metadata": {
    "ImageConfig": {
      "architecture": "",
      "created": "0001-01-01T00:00:00Z",
      "os": "",
      "rootfs": {
        "type": "",
        "diff_ids": null
      },
      "config": {}
    }
  },
  "Results": [
    {
      "Target": "Dockerfile",
      "Class": "config",
      "Type": "dockerfile",
      "MisconfSummary": {
        "Successes": 25,
        "Failures": 2,
        "Exceptions": 0
      },
      "Misconfigurations": [
        {
          "Type": "Dockerfile Security Check",
          "ID": "DS002",
          "AVDID": "AVD-DS-0002",
          "Title": "Image user should not be 'root'",
          "Description": "Running containers with 'root' user can lead to a container escape situation. It is a best practice to run containers as non-root users, which can be done by adding a 'USER' statement to the Dockerfile.",
          "Message": "Specify at least 1 USER command in Dockerfile with non-root user as argument",
          "Namespace": "builtin.dockerfile.DS002",
          "Query": "data.builtin.dockerfile.DS002.deny",
          "Resolution": "Add 'USER \u003cnon root user name\u003e' line to the Dockerfile",
          "Severity": "HIGH",
          "PrimaryURL": "https://avd.aquasec.com/misconfig/ds002",
          "References": [
            "https://docs.docker.com/develop/develop-images/dockerfile_best-practices/",
            "https://avd.aquasec.com/misconfig/ds002"
          ],
          "Status": "FAIL",
          "Layer": {},
          "CauseMetadata": {
            "Provider": "Dockerfile",
            "Service": "general",
            "Code": {
              "Lines": null
            }
          }
        },
        {
          "Type": "Dockerfile Security Check",
          "ID": "DS026",
          "AVDID": "AVD-DS-0026",
          "Title": "No HEALTHCHECK defined",
          "Description": "You should add HEALTHCHECK instruction in your docker container images to perform the health check on running containers.",
          "Message": "Add HEALTHCHECK instruction in your Dockerfile",
          "Namespace": "builtin.dockerfile.DS026",
          "Query": "data.builtin.dockerfile.DS026.deny",
          "Resolution": "Add HEALTHCHECK instruction in Dockerfile",
          "Severity": "LOW",
          "PrimaryURL": "https://avd.aquasec.com/misconfig/ds026",
          "References": [
            "https://blog.aquasec.com/docker-security-best-practices",
            "https://avd.aquasec.com/misconfig/ds026"
          ],
          "Status": "FAIL",
          "Layer": {},
          "CauseMetadata": {
            "Provider": "Dockerfile",
            "Service": "general",
            "Code": {
              "Lines": null
            }
          }
        }
      ]
    }
  ]
}
```

### Operating System

Ubuntu 24.04 LTS.  Trivy installed from nix unstable channel: 2yz2gcghw8pzbvhyihlls1m31ra1hn19-trivy-0.51.2

### Version

```bash
Version: v0.51.2
Check Bundle:
  Digest: sha256:6d0771effa53c6cf8130861fc3ac28f5515c35a028edb4bb1e67261b9218c80e
  DownloadedAt: 2024-05-23 03:59:36.949766554 +0000 UTC
```

### Checklist

- [X] Run `trivy image --reset`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by simar7** [comment]
> The cause is in /lua/lint/linters/trivy.lua. 

Is this part of the Trivy OSS project? Or did you mean to file this issue elsewhere?

**Reply by bobsrac** [reply]
Ha.  Yep.  Elsewhere :)

**Reply by bobsrac** [reply]
Bug was for the nvim-lint project.  Thanks for pointing that out.  Maybe it will actually get patched this way.
https://github.com/mfussenegger/nvim-lint/issues/588
